### PR TITLE
#325 small enhancement to fix the GetLinkFromUrl method

### LINKED ISSUE #325: Shaarli does not recognize saved links
Hi, I'm running the shaarli version 0.5.1 and when I save a link shaarli does not have problems. But if I want to save the same link again shaarli does not load the tags.. issue here :(

### PR DESCRIPTION
**Pull request by doc75**
I just move the fix #325 to fix the GetLinkFromUrl method.
It would avoid in the future that somebody is calling this method with an Url object instead of a string.
I also added a unit test to ensure that GetLinkFromUrl accept a Url object (and avoid further regression).

### CODE REVIEW

**Comment by virtualtam** [comment]
Hello @doc75!

I'm actually wondering whether it was a good idea to introduce the _not-so-magic_ `__toString()` method in the first place, as it seems to cause more indirect issues than the mere convenience it is expected to bring.

Casting variables to strings despite being by design supposed to be considered as such turns out to be quite developer-unfriendly...

**Comment by doc75** [comment]
@virtualtam: I think that the cast is more convenient in the callee than in the caller in any case.
Do you prefer that we add a method to_s() on Url object to make it more obvious and avoid the cast ?
To remove the __toString() I would let somebody more experienced in Shaarli code do it, as I am afraid to break everything ;-)

**Comment by virtualtam** [comment]
TL;DR an explicit call to a string conversion would be preferable

Well, I may have a little bias from working with Python, where an object's string representation (`my_object.__str__()` // `my_object.__unicode__()`) is called whenever needed - hence the will to use PHP's `__toString()`, which... does not behave as expected.

This being acknowledged, using semi-working tricks is very likely to entail undesired side-effects, the first ones being confusing developers and slowing bug seeking.

If you're willing to get familiar with Shaarli's codebase, moving from `__toString()` to an explicit `toString()` method should be quite easy, e.g.:
- update test code
- update `Url` code to get the tests passing
- update client calls in `index.php` - replace casts by explicit calls to `Url->toString()`

Also, there are plenty of resources within the Wiki's [Development](https://github.com/shaarli/Shaarli/wiki/Development) section ;-)

**Comment by doc75** [comment]
@virtualtam: I have removed __toString and it seems to work fine on my installation. Hope this is OK for you.

File: index.php [hunk 1]
```diff
@@ -1501,7 +1501,7 @@ function renderPage()
 
         $link_is_new = false;
         // Check if URL is not already in database (in this case, we will edit the existing link)
-        $link = $LINKSDB->getLinkFromUrl((string)$url);
```
Comment (ArthurHoaro) [inline index.php]: That's not enough. `$url` is used in the next lines, and expected to be 'auto-converted' as a string. Without `__toString()`, you need to explicitely call `toString()`.

Comment (virtualtam) [inline index.php]: question: is a `Url` object worth using in following operations?

if not, we could simply use string Urls from the beginning (l. 1499 / 1500)

File: tests/LinkDBTest.php [hunk 2]
```diff
@@ -7,6 +7,7 @@
 require_once 'application/LinkDB.php';
 require_once 'application/Utils.php';
 require_once 'tests/utils/ReferenceLinkDB.php';
+require_once 'application/Url.php';
```
Comment (virtualtam) [inline tests/LinkDBTest.php]: import sorting order:
- application imports
- test imports

File: application/LinkDB.php [hunk 3]
```diff
@@ -290,8 +290,14 @@ public function savedb($pageCacheDir)
      */
     public function getLinkFromUrl($url)
     {
-        if (isset($this->_urls[$url])) {
-            return $this->_links[$this->_urls[$url]];
+        // $url can be a string or an Url object
+        if( is_string($url) ) {
```
Comment (virtualtam) [inline application/LinkDB.php]: formatting convention:
```php
if (is_string($url)) {
    ...
}
```

maybe it can be handled with a `try/catch` block as well:
```php
try {
    $url_str = $url->toString();
} catch (WhateverMayBeRaised $exc) {
    $url_str = $url;
}
```

File: application/LinkDB.php [hunk 4]
```diff
@@ -290,8 +290,14 @@ public function savedb($pageCacheDir)
      */
     public function getLinkFromUrl($url)
     {
-        if (isset($this->_urls[$url])) {
-            return $this->_links[$this->_urls[$url]];
+        // $url can be a string or an Url object
```
Comment (virtualtam) [inline application/LinkDB.php]: this precision can be given in the method's parameter docstring (which, in this case, is missing ^^):
```php
/**
 * Returns the link for a given URL, or False if it does not exist.
 *
 * @param mixed $url URL to search for (either a string or a 'Url' object)
 *
 * @return the existing link if it exists, else 'false'
 */
```

**Comment by virtualtam** [comment]
Neat!
Thanks @doc75 for updating the PR, I'll take some time this evening to test it ;-)

File: tests/LinkDBTest.php (hunk 2 above)
Comment (doc75) [inline tests/LinkDBTest.php]: OK, re-ordered in next commit

File: application/LinkDB.php (hunk 3 above)
Comment (doc75) [inline application/LinkDB.php]: Formatting convention fixed in next commit.
Personnally I prefer the actual test which is more explicit. But let me know if this si really better to use exception.

File: application/LinkDB.php (hunk 4 above)
Comment (doc75) [inline application/LinkDB.php]: added in next commit, this is a good idea

File: index.php (hunk 1 above)
Comment (doc75) [inline index.php]: @ArthurHoaro: Good catch, I did it too fast.
I committed the necessary changes that I have seen and changed the GetHTTP method to handle both Url and string.

@virtualtam: I guess the Url class is to ease the usage of cleanup, but the question is valid.

Comment (virtualtam) [inline index.php]: > Url class is to ease the usage of cleanup

This was the whole purpose of the refactoring in #314

> the question is valid

there's a choice to make here, whether:
- all methods must support string- and object-represented URLs (some usages may be missed and regressions introduced)
- the Url class is used in a single place to perform URL cleanup and return a string, and all methods expect to receive their `$url` parameter as a string (reduces code impact -keep in mind that `index.php` is not test-covered)

File: tests/LinkDBTest.php [hunk 5]
```diff
@@ -6,6 +6,8 @@
 require_once 'application/Cache.php';
 require_once 'application/LinkDB.php';
 require_once 'application/Utils.php';
+require_once 'application/Url.php';
```
Comment (virtualtam) [inline tests/LinkDBTest.php]: for the sole sake of being a picky person: still unsorted :)

Comment (doc75) [inline tests/LinkDBTest.php]: No problem, this is a way to ensure proper code ;-)
PR updated

File: index.php (hunk 1 above)
Comment (doc75) [inline index.php]: Url also allow in 1 location to retrieve the scheme.
My 2 cents: I would create 2 methods cleanup_url and get_url_scheme th

Comment (virtualtam) [inline index.php]: +1 for keeping Url as-is and introduce a `cleanup_url` helper function

Regarding `getHTTP`, it's already undergoing changes in:
- #333 -  Refactor HTTP / Url functions
    - #334 - HTTP: move utils to a proper file, add tests (WIP)
    - _<insert future related PRs>_

**Comment by doc75** [comment]
@virtualtam: I created 2 static methods in Url to allow cleanup and getScheme. I moved the original public method to protected.
If this is fine for you let me know, I will test further this PR.

**Comment by virtualtam** [comment]
Hey @doc75, thanks for updating the PR :)

The helper functions should preferably not be static members of the `Url` class; it brings a bit of confusion to call a static `Url::cleanupUrl()` that itself instantiates a `Url` and then calls `Url->cleanup()`.

A cleaner solution would be to have the `Url` class on the one hand, and a couple global functions on the other hand:

```php
class Url
{
    public function someMethod($someParam)
    {
        doSomeStuff();
    }
}

function some_helper_function($someParam)
{
    $url = new Url();
    return $url->someMethod();
}
```

Also notice that the naming differs from global functions to class members, we settled on a Zend/PSR convention, see the [example](https://github.com/shaarli/Shaarli/issues/95#issuecomment-120890850) in #95

### Recap'
To sum up, you'd end with the following code:

LinkDB:
- no change, expect string-formatted URLs (no if/else type check)

Url class:
- rename `__toString` as `toString`
- update test code to explicitely call `toString` instead of implicit casting

Url helper function:
- cleanup_url()
- add a new test class to cover it ;-)

index.php:
- call the helper function to cleanup the URL and return a string
- look for the scheme in the string-represented URL

### Git usage
To keep history tidy while working on a PR, there are several possibilities:
1. commit sandwich (useful if the feature requires several steps befor being implemented)
    - create a dev-branch
    - submit a first commit to create a PR
    - take reviews into account by adding new commits to the dev-branch
    - once everyone is okay with the changes, [squash](http://www.git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Squashing-Commits) all commits into a single one
2. keep a unique commit up-to-date (preferred)
    - create a dev-branch
    - submit a first commit to create a PR
    - [amend](http://www.git-scm.com/book/en/v2/Git-Basics-Undoing-Things) the commit to update its contents
    - force-push the dev-branch (rewrites its history - dev-branches are the only place where this happens)

When the mainline (master branch) changes in such a way that your commit(s) create conflicts, you should:
- [rebase](http://www.git-scm.com/docs/git-rebase) the dev-branch upon the mainline
- resolve conflicts
- re-submit the dev-branch

Feel free to PM me on [Gitter](https://gitter.im/shaarli/Shaarli) if you need a hand with the changes ;-)

**Comment by doc75** [comment]
@virtualtam I think this is pretty clear. Just one point in the helper function, should it go in application/Utils.php ? I think that it should not be in index.php to be testable.
I 'll then follow your recommendation.

**Comment by virtualtam** [comment]
`application/Url.php` sounds like a good place, whereas `application/Utils.php` is rather for generic functions

**Comment by doc75** [comment]
Another idea, why not keeping the static functions as helpers and move all other functions as protected (including the __construct function), this would prevent further easy usage of Url object by other developers.
And thanks for your patience.

**Comment by virtualtam** [comment]
The issue here is that there's a lot of cleanup to do in `index.php` and no one currently has a full-understanding of Shaarli's internals; we're learning by trial & through successive refactorings that highlight the code's logic: function calls, object interactions, possible dependency injections, dead instructions, etc .

So maybe in the near future we'll figure it's more convenient to use the `Url` class and will be confident enough to propagate it everywhere...

There would very little benefit in changing the scope of `Url`'s members, and it would force rewriting all unitary tests in a very inconvenient way :)

True, a developer may be lured into using it instead of a helper function... but we've got code reviews to prevent this :)

**Comment by doc75** [comment]
Fair enough, I'll follow the directions provided in your previous comment and let you know when this is done.

**Comment by doc75** [comment]
@virtualtam: I updated the PR according to your advices. Let me know if I missed anything.

File: tests/UrlTest.php [hunk 6]
```diff
@@ -30,6 +30,109 @@ public function testUnparseFull()
...
+    {
+        $this->assertEquals('', get_url_scheme(''));
+    }
+
+    /**
+     * Get normal scheme of Url
+     */
+    public function testGetUrlScheme()
+    {
+        $url = 'http://domain.tld:3000';
```
Comment (virtualtam) [inline tests/UrlTest.php]: no need to assign `$url`, you can directly call:
```php
this->assertEquals(
    'scheme',
    get_url_scheme('scheme://test.url')
);
```

File: tests/UrlTest.php [hunk 7]
```diff
@@ -30,6 +30,109 @@ public function testUnparseFull()
...
+        $this->assertEquals('http', get_url_scheme($url));
+        $url = 'https://domain.tld:3000';
+        $this->assertEquals('https', get_url_scheme($url));
+        $url = 'domain.tld';
+        $this->assertEquals('http', get_url_shceme($url));
+        $url = 'ssh://domain.tld';
+        $this->assertEquals('ssh', get_url_shceme($url));
+        $url = 'ftp://domain.tld';
+        $this->assertEquals('ftp', get_url_shceme($url));
```
Comment (virtualtam) [inline tests/UrlTest.php]: tests should return an error when looking for `get_url_shceme` (typo: `shceme`)

they can be run locally using the Makefile, see [Unit tests](https://github.com/shaarli/Shaarli/wiki/Unit-tests) on the wiki :)

Travis did not run a build for this commit, it happens from time to time... grr

**Comment by virtualtam** [comment]
Almost good! :+1: 
I Ieft comments for a couple minor aspects (test typos)

File: tests/UrlTest.php (hunk 7 above)
Comment (doc75) [inline tests/UrlTest.php]: I have seen that when looking at your previous comment.
Indeed when running the test on my machine with `vendor/bin/phpunit tests/UrlTest.php it runs only the first class in the file (2 tests).
I am quite unexperienced with phpunit (and php...), is this a problem on my side or common phpunit behavior (which would mean that only 1 class per test file is needed) ?

Comment (doc75) [inline tests/UrlTest.php]: @virtualtam: After some search on internet it appears to be "normal". I'll split the test and check if there are no other file with the same mistake. Let me know if I am wrong.

Comment (virtualtam) [inline tests/UrlTest.php]: - there can be several (test) classes per source files
    - but there cannot be two with the same name, unless using namespaces (which is not the case for Shaarli yet)
- when Xdebug is installed and enabled, an HTML coverage report will be generated, that can be displayed by opening `coverage\index.html` in a web browser
- code sections interpreted during a test suite will appear as _covered_, whereas unevaluated sections will appear as _uncovered_

**Comment by doc75** [comment]
I tested with Xdebug and can confirm that on php 5.3 at least putting several class in the same file leads to replay of only the first class. I therefore splitted each class in a file to ensure they are all run.
It should be OK now...
Let me know.

**Comment by virtualtam** [comment]
Checked as well and you're perfectly right, quite a good catch on this one :)

**Comment by virtualtam** [comment]
Tested, works fine!

Last request: could you update the commit message as follows?

```
Url: introduce global helper functions for cleanup and scheme detection

Relates to #314 & #326

Additions:
 - add global `cleanup_url()` and `get_url_scheme()` functions

Modifications:
 - replace `Url` usage in `index.php` by calls to global functions
 - fix `Url` tests not being run: PHPUnit expects a single test class per file
   - move classes to separate files
```

**Comment by doc75** [comment]
Commit message updated

**Comment by virtualtam** [comment]
Merged, many thanks for your patience :)

**Comment by doc75** [comment]
Thanks for the merge and your patience also.
